hnn_core/parallel_backends.py

```diff
@@ -137,23 +146,27 @@ def run_subprocess(command, obj, timeout, *args, **kwargs):
         threads_started = True
         data_received = False
         sent_network = False
+        count_since_last_output = 0
 
         # loop while the process is running the simulation
         while True:
             child_terminated = proc.poll() is not None
 
             if not data_received:
+                if _echo_child_output(out_q):
+                    count_since_last_output = 0
+                else:
+                    count_since_last_output += 1
                 # look for data in stderr and print child stdout
-                data_len, proc_data_bytes = _get_data_from_output(out_q, err_q)
+                data_len, proc_data_bytes = _get_data_from_child_err(err_q)
                 if data_len > 0:
                     data_received = True
                     _write_child_exit_signal(proc.stdin)
                 elif child_terminated:
-                    # child terminated early, but we ran
-                    # _get_data_from_output() above to capture
-                    # any output that was left in queues
-                    warn("Child process failed unexpectedly. Output from"
-                         " child below:")
+                    # child terminated early, and we already
+                    # captured output left in queues
+                    warn("Child process failed unexpectedly")
+                    kill_proc_name('nrniv')
```

Comment:
wouldn't this be called in `__exit__` anyway?

Reply:
Yes, but it facilitates collecting errors from the child process at `proc.communiate()`. If we wait until `__exit__`, we won't be checking for output from the child.

Reply:
It also has to do with MPI-failures vs. python values. Killing self.proc stops `mpi_child.py`, but the nrniv processes may linger in the background. Killing them explicitly makes sure that output is flushed and `mpi_child.py` has an opportunity to relay those messages to `MPIBackend`

Reply:
I see, got it. Thanks for explaining!